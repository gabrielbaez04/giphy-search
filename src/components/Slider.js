import React from 'react';
import RightArrow from './RightArrow';
import LeftArrow from './LeftArrow';
import Slide from './Slide';
import PropTypes from 'prop-types';

class Slider extends React.Component {
  state = {
    currentIndex: this.props.selectedGif || 0,
    translateValue: this.props.selectedGif > 0 ? -500 * this.props.selectedGif : 0
  }

  goToPrevSlide = () => {
    if (this.state.currentIndex === 0)
      return;

    this.setState(prevState => ({
      currentIndex: prevState.currentIndex - 1,
      translateValue: prevState.translateValue + this.slideWidth()
    }))
  }

  goToNextSlide = () => {
    if (this.state.currentIndex === this.props.gifs.length - 1) {
      return this.setState({
        currentIndex: 0,
        translateValue: 0
      })
    }

    // This will not run if we met the if condition above
    this.setState(prevState => ({
      currentIndex: prevState.currentIndex + 1,
      translateValue: prevState.translateValue + -(this.slideWidth())
    }));
  }

  slideWidth = () => {
    return document.querySelector('.slide').clientWidth
  }

  render() {
    return (
      <div className="slider">

        <div className="slider-wrapper"
          style={{
            transform: `translateX(${this.state.translateValue}px)`,
            transition: 'transform ease-out 0.45s'
          }}>
          {
            this.props.gifs.map((gif, i) => (
              <Slide key={i} image={gif.images.downsized.url} />
            ))
          }
        </div>

        <LeftArrow
          goToPrevSlide={this.goToPrevSlide}
        />

        <RightArrow
          goToNextSlide={this.goToNextSlide}
        />
      </div>
    );
  }
}

Slider.propTypes = {
  gifs: PropTypes.array.isRequired,
  selectedGif: PropTypes.number,
}

export default Slider;
