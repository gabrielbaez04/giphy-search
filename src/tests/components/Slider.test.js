import { shallow } from 'enzyme';
import Slider from '../../components/Slider';
import React from 'react';

let gifs, selectedGif;
beforeEach(() => {
    gifs = [
        { id: 1, title: "test", images: { downsized_still: { url: "test.url.com" }, downsized: { url: "test" } } },
        { id: 2, title: "test", images: { downsized_still: { url: "test.url.com" }, downsized: { url: "test" } } },
        { id: 3, title: "test", images: { downsized_still: { url: "test.url.com" }, downsized: { url: "test" } } },
    ];
    selectedGif = 0;
});

test('should render Search correctly', () => {
    const wrapper = shallow(<Slider gifs={gifs} selectedGif={selectedGif} />);
    expect(wrapper).toMatchSnapshot();
});
