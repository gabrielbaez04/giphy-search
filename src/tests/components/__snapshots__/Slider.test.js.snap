// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Search correctly 1`] = `
<div
  className="slider"
>
  <div
    className="slider-wrapper"
    style={
      Object {
        "transform": "translateX(0px)",
        "transition": "transform ease-out 0.45s",
      }
    }
  >
    <Slide
      image="test"
      key="0"
    />
    <Slide
      image="test"
      key="1"
    />
    <Slide
      image="test"
      key="2"
    />
  </div>
  <LeftArrow
    goToPrevSlide={[Function]}
  />
  <RightArrow
    goToNextSlide={[Function]}
  />
</div>
`;
